supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,589 @@
+
+        Returns:
+            A dictionary of the statuses of any jobs that have not been successfully completed.
+        """
+        # If no jobs then return empty dictionary
+        if not [sample for sample in self.samples if sample.job is not None]:
+            return {}
+
+        statuses = {}
+        waiting_samples = [
+            sample for sample in self.samples if not hasattr(sample, "probabilities")
+        ]
+        for sample in tqdm(waiting_samples, "Retrieving jobs"):
+            if sample.job is None:
+                continue
+            if (
+                job_status := sample.job.status()
+            ) in css.job.Job.NON_TERMINAL_STATES + css.job.Job.UNSUCCESSFUL_STATES:
+                statuses[sample.job.job_id()] = job_status
+            else:
+                sample.probabilities = self._process_device_counts(sample.job.counts(0))
+
+        return statuses
+
+    def _run_check(self) -> None:
+        """Checks if any of the samples already have probabilities stored. If so raises a runtime
+        error to prevent them from being overwritten.
+
+        To be used within all `run` methods to prevent data being overwritten
+
+        Raises:
+            RuntimeError: If any samples already have probabilities stored.
+        """
+        if any(hasattr(sample, "probabilities") for sample in self.samples):
+            raise RuntimeError(
+                "Some samples have already been run. Re-running the experiment will"
+                "overwrite these results. If this is the desired behaviour use `overwrite=True`"
+            )
+
+    def _sample_statuses(self) -> list[str | None]:
+        """Returns:
+        The statuses of the jobs associated with each sample. If no job is associated
+        with a sample then :code:`None` is listed instead.
+        """
+        statuses: list[str | None] = []
+        for sample in self.samples:
+            if sample.job is None:
+                statuses.append(None)
+            else:
+                statuses.append(sample.job.status())
+        return statuses
+
+    def _validate_circuits(self) -> None:
+        """Checks that all circuits contain a terminal measurement of all qubits."""
+        for sample in self.samples:
+            if not sample.circuit.has_measurements():
+                raise ValueError("QCVV experiment circuits must contain measurements.")
+            if not sample.circuit.are_all_measurements_terminal():
+                raise ValueError("QCVV experiment circuits can only contain terminal measurements.")
+            if not sorted(sample.circuit[-1].qubits) == sorted(self.qubits):
+                raise ValueError(
+                    "The terminal measurement in QCVV experiment circuits must measure all qubits."
+                )
+
+    ###################
+    # Public Methods  #
+    ###################
+    @abstractmethod
+    def analyse_results(self, plot_results: bool = True) -> ResultsT:
```

Comment:
kinda silly/low priority question: do we want to enforce any consistency re: UK vs US spellings?